sui/manage-coins.js

```diff
@@ -0,0 +1,218 @@
+
+class CoinManager {
+
+    static async getAllCoins(client, account) {
+        let cursor;
+        const coinTypeToCoins = {};
+
+        do {
+            const coinsAtCursor = await client.getAllCoins({
+                owner: account,
+                limit: 100,
+                cursor,
+            });
+
+            for (const coin of coinsAtCursor.data) {
+                if (!coinTypeToCoins[coin.coinType]) {
+                    coinTypeToCoins[coin.coinType] = {
+                        data: [],
+                        totalBalance: 0n,
+                    };
+                }
+
+                coinTypeToCoins[coin.coinType].data.push(coin);
+                coinTypeToCoins[coin.coinType].totalBalance += BigInt(coin.balance);
+            }
+
+            if (coinsAtCursor.hasNextPage) {
+                cursor = coinsAtCursor.nextCursor;
+            }
+        } while (cursor);
+
+        return coinTypeToCoins;
+    }
+
+    static async splitCoins(tx, coinTypeToCoins, options) {
+        CoinManager.checkSplitAmount(options);
+        const splitAmount = BigInt(options.split);
+
+        console.log('\n==== Splitting Coins ====');
+
+        // Set coin type to given coin type or the first coin type if there's only one if it's a SUI token.
+        const hasOnlyGasToken = Object.keys(coinTypeToCoins).length === 1 && coinTypeToCoins[SUI_COIN_ID];
+        const coinType = options.coinType || hasOnlyGasToken ? SUI_COIN_ID : undefined;
+
+        if (coinType) {
+            // Throw an error if the coin type is specified but no coins are found
+            CoinManager.checkCoinType(coinType, coinTypeToCoins);
+            const coins = coinTypeToCoins[coinType];
+            const [coin] = CoinManager.doSplitCoins(tx, coins, splitAmount);
+
+            if (options.transfer) {
+                CoinManager.doTransfer(tx, coin, options.transfer);
+            }
+        } else {
+            for (const coinType in coinTypeToCoins) {
+                const coins = coinTypeToCoins[coinType];
+                if (this.isGasToken(coins.data[0])) continue;
+                const [coin] = CoinManager.doSplitCoins(tx, coins, splitAmount);
+
+                if (options.transfer) {
+                    CoinManager.doTransfer(tx, coin, options.transfer);
+                }
+            }
+        }
+
+        if (options.transfer) {
+            console.log(`\nTransfer ${splitAmount} coins for every split coin to ${chalk.green(options.transfer)}`);
+        }
+
+        // The transaction will fail if the gas budget is not set for splitting coins transaction
+        tx.setGasBudget(1e8);
+    }
+
+    static doTransfer(tx, coin, recipient) {
+        tx.transferObjects([coin], recipient);
+    }
+
+    static doSplitCoins(tx, coins, splitAmount) {
+        const firstObjectId = this.isGasToken(coins.data[0]) ? tx.gas : coins.data[0].coinObjectId;
+        const response = tx.splitCoins(firstObjectId, [splitAmount]);
```

Comment:
splitAmount should be decimal aware. Script user will provide 1 as the amount to mean 1 SUI. does the Sui SDK handle this? If not, we should do it ourselves (by retrieving coin metadata)

Reply:
can't find it. so i'll do it manually using sui sdk to get decimals from coin metadata and use ethers.js to handle amount conversion.